### ISSUE TITLE
The way self.embed_avg is being computed

### ISSUE THREAD

**Issue by AndreyBocharnikov** [lucidrains/vector-quantize-pytorch#38]
In the current implementation the `self.embed_avg` is being computed as follows  ([code](https://github.com/lucidrains/vector-quantize-pytorch/blob/master/vector_quantize_pytorch/vector_quantize_pytorch.py#L300))
`embed_avg = ema(embed_avg, embed_sum) / (laplace_smoothing(self.cluster_size) * self.cluster_size.sum())`, where `embed_sum` is a tensor of shape `[n_classes, hidden_size]` and all `batch x time` samples of current batch that belong to the same class are being summed up. Since we want a mean, not a sum, we divide by `self.cluster_size`, but since some values might be equal to 0, we introduce the laplace_smoothing, that in its turn introduces the divition by `self.cluster_size.sum()` which we invalidate by further multiplying by it.

My question is, would not it be better to compute `embed_avg` as `embed_avg = ema(embed_avg,                       
 torch.where(self.cluster_size == 0, 0, embed_sum / self.cluster_size.unsqueeze(dim=-1)))` It seems more accurate since we divide `embed_sum`, not the `ema(embed_avg, embed_sum)` by `cluster_size` and there are no epsilons that are appear in both numerator and denominator. 

Maybe I just did not get some intuition behind your code, could you tell me please, why did you write it the way it is? Is it referring to an algorithm from some paper? Or have you taken this code from some other repository?

**Comment by lucidrains** [comment]
@AndreyBocharnikov hey Andrey, indeed it was faithfully inherited from the original deepmind sonnet implementation [here](https://github.com/deepmind/sonnet/blob/v2/sonnet/src/nets/vqvae.py#L270)

**Comment by lucidrains** [comment]
![giphy](https://user-images.githubusercontent.com/108653/218209377-be90b1e0-d0bc-4460-8547-917a92a38093.gif)
